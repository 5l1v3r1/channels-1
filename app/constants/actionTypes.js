// @flow
const actionTypes = {
  FETCH_ACCESS_TOKEN_REQUEST: 'FETCH_ACCESS_TOKEN_REQUEST',
  FETCH_ACCESS_TOKEN_SUCCESS: 'FETCH_ACCESS_TOKEN_SUCCESS',
  FETCH_ACCESS_TOKEN_FAILURE: 'FETCH_ACCESS_TOKEN_FAILURE',
  FETCH_CHANNELS_REQUEST: 'FETCH_CHANNELS_REQUEST',
  FETCH_CHANNELS_SUCCESS: 'FETCH_CHANNELS_SUCCESS',
  FETCH_CHANNELS_FAILURE: 'FETCH_CHANNELS_FAILURE',
  UPDATE_ACTIVE_CHANNEL: 'UPDATE_ACTIVE_CHANNEL',
  FETCH_VIDEOS_REQUEST: 'FETCH_VIDEOS_REQUEST',
  FETCH_VIDEOS_SUCCESS: 'FETCH_VIDEOS_SUCCESS',
  FETCH_VIDEOS_FAILURE: 'FETCH_VIDEOS_FAILURE'
};

export default actionTypes;
