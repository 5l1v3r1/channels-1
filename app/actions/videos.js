// @flow
import actionTypes from '../constants/actionTypes';
import type { Action, ThunkAction, Dispatch, GetState } from '../constants/types';
import { IVideo } from '../dataTypes/videoType';
import { getRequest } from '../services/fetch';
import { getVideosUri, getVideoIdsUri } from '../services/uriGenerator';

export function fetchVideos(): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    dispatch(fetchVideosRequest());
    const params = {
      channelId: getState().channels.activeId,
      pageToken: getState().videos.pageToken,
      order: 'date'
    };
    return fetchVideoIds(params)
      .then(
        res => {
          const { videoIds, nextPageToken } = res;
          const uri = getVideosUri(videoIds);
          return getRequest(uri)
            .then(
              json => dispatch(fetchVideosSuccess(json.items, nextPageToken)),
              () => dispatch(fetchVideosFailure())
            );
        },
        () => dispatch(fetchVideosFailure())
      );
  };
}

export function fetchVideoIds(params: any): Promise<any> {
  const uri = getVideoIdsUri(params);
  return getRequest(uri)
    .then(
      json => Promise.resolve({
        videoIds: json.items.map(item => item.id.videoId).join(),
        nextPageToken: json.nextPageToken || params.pageToken
      }),
      () => Promise.reject()
    );
}

export function fetchVideosRequest(): Action {
  return {
    type: actionTypes.FETCH_VIDEOS_REQUEST
  };
}

export function fetchVideosSuccess(items: IVideo[], nextPageToken: string): Action {
  return {
    type: actionTypes.FETCH_VIDEOS_SUCCESS,
    payload: {
      items,
      nextPageToken
    }
  };
}

export function fetchVideosFailure(): Action {
  return {
    type: actionTypes.FETCH_VIDEOS_FAILURE
  };
}
