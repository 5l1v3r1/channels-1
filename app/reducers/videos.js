// @flow
import { combineReducers } from 'redux';
import actionTypes from '../constants/actionTypes';
import type { Action, VideosState } from '../constants/typeAliases';
import VideoType from '../dataTypes/videoType';
import { stringOrEmpty } from '../utils/utils';

function byId(state: { [string]: VideoType } = {}, { type, payload }: Action): { [string]: VideoType } {
  switch (type) {
    case actionTypes.FETCH_VIDEOS_SUCCESS:
      const nextState = { ...state };
      payload.items.forEach(item => {
        nextState[item.id] = VideoType.from(item);
      });
      return nextState;
    case actionTypes.UPDATE_ACTIVE_CHANNEL:
      return {};
    default:
      return state;
  }
}

function allIds(state: string[] = [], { type, payload }: Action): string[] {
  switch (type) {
    case actionTypes.FETCH_VIDEOS_SUCCESS:
      return [
        ...state,
        ...payload.items.map(item => stringOrEmpty(item.id))
      ];
    case actionTypes.UPDATE_ACTIVE_CHANNEL:
      return [];
    default:
      return state;
  }
}

function pageToken(state: string = '', { type, payload }: Action): string {
  switch (type) {
    case actionTypes.FETCH_VIDEOS_SUCCESS:
      return payload.nextPageToken;
    case actionTypes.UPDATE_ACTIVE_CHANNEL:
      return '';
    default:
      return state;
  }
}

function isFetching(state: boolean = false, { type }: Action): boolean {
  switch (type) {
    case actionTypes.FETCH_VIDEOS_REQUEST:
      return true;
    case actionTypes.FETCH_VIDEOS_SUCCESS:
    case actionTypes.FETCH_VIDEOS_FAILURE:
      return false;
    default:
      return state;
  }
}

function error(state: boolean = false, { type }: Action): boolean {
  switch (type) {
    case actionTypes.FETCH_VIDEOS_REQUEST:
    case actionTypes.FETCH_VIDEOS_SUCCESS:
      return false;
    case actionTypes.FETCH_VIDEOS_FAILURE:
      return true;
    default:
      return state;
  }
}

const videos = combineReducers({
  byId,
  allIds,
  pageToken,
  isFetching,
  error
});

export function getVideos(state: VideosState): VideoType[] {
  return state.allIds.map(id => state.byId[id]);
}

export default videos;
