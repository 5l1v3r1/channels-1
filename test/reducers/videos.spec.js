import videos from '../../app/reducers/videos';
import actionTypes from '../../app/constants/actionTypes';
import VideoType from '../../app/dataTypes/videoType';

describe('videos reducer', () => {
  it('should return initial state', () => {
    expect(videos(undefined, {})).toEqual({
      byId: {},
      allIds: [],
      pageToken: '',
      isFetching: false,
      error: false
    });
  });

  it('should handle FETCH_VIDEOS_REQUEST', () => {
    const action = { type: actionTypes.FETCH_VIDEOS_REQUEST };
    expect(videos(undefined, action)).toEqual({
      byId: {},
      allIds: [],
      pageToken: '',
      isFetching: true,
      error: false
    });
  });

  it('should handle FETCH_VIDEOS_SUCCESS', () => {
    VideoType.from = jest.fn(item => item);

    expect((
      videos(
        {
          byId: {},
          allIds: [],
          pageToken: '',
          isFetching: true,
          error: false
        },
        {
          type: actionTypes.FETCH_VIDEOS_SUCCESS,
          payload: {
            items: [
              { id: 'M8l2aGMjKHI' },
              { id: 'GcSACxUbqtg' }
            ],
            nextPageToken: 'CBQQAA'
          }
        }
      )
    )).toEqual({
      byId: {
        M8l2aGMjKHI: {
          id: 'M8l2aGMjKHI'
        },
        GcSACxUbqtg: {
          id: 'GcSACxUbqtg'
        }
      },
      allIds: [
        'M8l2aGMjKHI',
        'GcSACxUbqtg'
      ],
      pageToken: 'CBQQAA',
      isFetching: false,
      error: false
    });

    expect((
      videos(
        {
          byId: {
            M8l2aGMjKHI: {
              id: 'M8l2aGMjKHI'
            },
            GcSACxUbqtg: {
              id: 'GcSACxUbqtg'
            }
          },
          allIds: [
            'M8l2aGMjKHI',
            'GcSACxUbqtg'
          ],
          pageToken: 'CBQQAA',
          isFetching: true,
          error: false
        },
        {
          type: actionTypes.FETCH_VIDEOS_SUCCESS,
          payload: {
            items: [
              { id: 'Flze-rwT7lM' },
              { id: 'L4TIM6W7u-M' }
            ],
            nextPageToken: 'CCgQAA'
          }
        }
      )
    )).toEqual({
      byId: {
        M8l2aGMjKHI: {
          id: 'M8l2aGMjKHI'
        },
        GcSACxUbqtg: {
          id: 'GcSACxUbqtg'
        },
        'Flze-rwT7lM': {
          id: 'Flze-rwT7lM'
        },
        'L4TIM6W7u-M': {
          id: 'L4TIM6W7u-M'
        }
      },
      allIds: [
        'M8l2aGMjKHI',
        'GcSACxUbqtg',
        'Flze-rwT7lM',
        'L4TIM6W7u-M'
      ],
      pageToken: 'CCgQAA',
      isFetching: false,
      error: false
    });
  });

  it('should handle FETCH_VIDEOS_FAILURE', () => {
    const state = {
      byId: {},
      allIds: [],
      pageToken: '',
      isFetching: true,
      error: false
    };
    const action = { type: actionTypes.FETCH_VIDEOS_FAILURE };
    expect(videos(state, action)).toEqual({
      byId: {},
      allIds: [],
      pageToken: '',
      isFetching: false,
      error: true
    });
  });

  it('should handle UPDATE_ACTIVE_CHANNEL', () => {
    const state = {
      byId: {
        M8l2aGMjKHI: {
          id: 'M8l2aGMjKHI'
        },
        GcSACxUbqtg: {
          id: 'GcSACxUbqtg'
        }
      },
      allIds: [
        'M8l2aGMjKHI',
        'GcSACxUbqtg'
      ],
      pageToken: 'CBQQAA',
      isFetching: false,
      error: false
    };
    const action = {
      type: actionTypes.UPDATE_ACTIVE_CHANNEL,
      payload: 'UCO1cgjhGzsSYb1rsB4bFe4Q'
    };
    expect(videos(state, action)).toEqual({
      byId: {},
      allIds: [],
      pageToken: '',
      isFetching: false,
      error: false
    });
  });

  it('should handle unknown action type', () => {
    const action = { type: 'unknown' };
    expect(videos(undefined, action)).toEqual({
      byId: {},
      allIds: [],
      pageToken: '',
      isFetching: false,
      error: false
    });
  });
});
