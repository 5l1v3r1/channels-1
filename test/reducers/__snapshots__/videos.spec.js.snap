// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`video reducer should handle FETCH_VIDEOS_FAILURE 1`] = `
Object {
  "allIds": Array [],
  "byId": Object {},
  "error": true,
  "isFetching": false,
  "pageToken": "",
}
`;

exports[`video reducer should handle FETCH_VIDEOS_REQUEST 1`] = `
Object {
  "allIds": Array [],
  "byId": Object {},
  "error": false,
  "isFetching": true,
  "pageToken": "",
}
`;

exports[`video reducer should handle FETCH_VIDEOS_SUCCESS 1`] = `
Object {
  "allIds": Array [
    "UCyIe-61Y8C4_o-zZCtO4ETQ",
    "UCO1cgjhGzsSYb1rsB4bFe4Q",
  ],
  "byId": Object {
    "UCO1cgjhGzsSYb1rsB4bFe4Q": Object {
      "id": "UCO1cgjhGzsSYb1rsB4bFe4Q",
    },
    "UCyIe-61Y8C4_o-zZCtO4ETQ": Object {
      "id": "UCyIe-61Y8C4_o-zZCtO4ETQ",
    },
  },
  "error": false,
  "isFetching": false,
  "pageToken": "CBQQAA",
}
`;

exports[`video reducer should handle UPDATE_ACTIVE_CHANNEL 1`] = `
Object {
  "allIds": Array [],
  "byId": Object {},
  "error": false,
  "isFetching": false,
  "pageToken": "",
}
`;

exports[`video reducer should handle unknown action type 1`] = `
Object {
  "allIds": Array [],
  "byId": Object {},
  "error": false,
  "isFetching": false,
  "pageToken": "",
}
`;

exports[`video reducer should return initial state 1`] = `
Object {
  "allIds": Array [],
  "byId": Object {},
  "error": false,
  "isFetching": false,
  "pageToken": "",
}
`;
